Feature/dx 638 3ds2 (#518)
SwedbankPay/developer.swedbankpay.com: _includes/payment-order-other-features.md

| {% icon check %} | └─➔&nbsp;`language`            | `string`     | {% include field-description-language.md api_resource="paymentorders" %}                                                                                                                                                                                                              |
| {% icon check %} | └─➔&nbsp;`urls`                | `string`     | {he URI to the `urls` resource where all URIs related to the payment order can be retrieved.                                                                                                                                                                                          |
| {% icon check %} | └─➔&nbsp;`callbackUrl`         | `string`     | The URL that Swedbank Pay will perform an HTTP `POST` against every time a transaction is created on the payment. See [callback][callback-reference] for details.                                                                                                                     |

> | {% icon check %} | └➔&nbsp;`payeeInfo`            | `string`     | The `payeeInfo` object, containing information about the payee (the recipient of the money).                                                                                                                                                                                          |

Reviewer: ```suggestion
| {% icon check %} | └➔&nbsp;`payeeInfo`            | `object`     | The `payeeInfo` object, containing information about the payee (the recipient of the money).                                                                                                                                                                                          |
```

| `paymentorder`            | `string` | {% include field-description-id.md resource="paymentorder" sub_resource="payer" %} |
| `payer`                   | `object` | The payer object.                                                                  |
| └➔&nbsp;`id`              | `string` | {% include field-description-id.md resource="payer" %}                             |

> | └➔&nbsp;`email`           | `string` | Payer's registered email address. The field is related to [3D Secure 2.0](/{{ documentation_section }}/other-features#3d-secure-20).   |

Reviewer: ```suggestion
| └➔&nbsp;`email`           | `string` | Payer's registered email address. The field is related to [3-D Secure 2](/{{ documentation_section }}/other-features#3d-secure-20).   |
```

| └➔&nbsp;`msisdn`          | `string` | Payer's registered mobile phone number.

> The field is related to [3D Secure 2.0](/{{ documentation_section }}/other-features#3d-secure-20).|

Reviewer: ```suggestion
| └➔&nbsp;`msisdn`          | `string` | Payer's registered mobile phone number. The field is related to [3-D Secure 2](/{{ documentation_section }}/other-features#3d-secure-20).|
```

{% include settlement-reconciliation.md api_resource="paymentorders" %}

> ## 3D Secure 2.0

Reviewer: ```suggestion
## 3-D Secure 2
```

When dealing with card payments, 3D Secure authentication of the
cardholder is an essential topic. 3D Secure 2.0 is an improved version of the
old protocol, now allowing frictionless payments where transactions can be
completed without input from the cardholder. Therefore, there are certain fields

> that should be included when implementing 3D Secure 2.0. These are listed below: 

Reviewer: ```suggestion
When dealing with card payments, 3-D Secure authentication of the cardholder is
an essential topic. 3-D Secure 2.0 is an improved version of the 3-D Secure
protocol, now allowing frictionless payments where transactions can be completed
without input from the cardholder. Therefore, there are certain fields that
should be included to allow frictionless 3-D Secure 2.0. These fields are listed
below: 

```

> *  `payment.riskIndicator`

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `payment.riskIndicator`

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `payment.cardholder`

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `payment.cardholder`

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *   `payment.cardholder`

Reviewer: Are the `payment` fields relevant for payment order?

> *  `paymentorder.payer.email` 

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `paymentorder.payer.email` 

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `paymentorder.payer.msisdn`

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `paymentorder.payer.msisdn`

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `paymentorder.payer.workPhoneNumber` 

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `paymentorder.payer.workPhoneNumber` 

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `paymentorder.payer.homePhoneNumber`

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *  `paymentorder.payer.homePhoneNumber`

Reviewer 2: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5786122)

> *   `paymentorder.payer.homePhoneNumber`

Reviewer: Perhaps we should enlist all fields within `payer` by just providing a full JSON code example?

understood in any response **must be ignored**. Don't expect a specific order of
elements. When in doubt, please follow the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle)." %}

> 

Reviewer: ```suggestion
```

{% include settlement-reconciliation.md api_resource="paymentorders" %}

## 3-D Secure 2

When dealing with card payments, 3-D Secure authentication of the
cardholder is an essential topic. 3-D Secure 2 is an improved version of the
old protocol, now allowing frictionless payments where transactions can be
completed without input from the cardholder. Therefore, there are certain fields
that should be included when implementing 3-D Secure 2. These are listed below
in the abbreviated `Purchase` request. 

{:.code-header}
**Request**

```http
POST /psp/paymentorders HTTP/1.1
Host: {{ page.api_host }}
Authorization: Bearer <AccessToken>
Content-Type: application/json

{
    "paymentorder": {
        "operation": "Purchase",
        "currency": "SEK",
        "amount": 1500,
        "vatAmount": 375,
        "description": "Test Purchase",
        "generatePaymentToken": true,
        "userAgent": "Mozilla/5.0...",
        "urls": {
            "hostUrls": [ "https://example.com", "https://example.net" ]
        },
        "payer": {
            "email": "olivia.nyhuus@payex.com",
            "msisdn": "+4798765432",
            "workPhoneNumber" : "+4787654321",

>             "homePhoneNumber" : "+4776543210"

Reviewer: Can't `shippingAddress` or what it's called also be sent in here?

        }
    }
}
```

{:.table .table-striped}
|     Required     | Field                             | Type         | Description                                                                                                                                                                                                                                                                                                                                                                                                   |
| :--------------: | :-------------------------------- | :----------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> |                  | └─➔&nbsp;`email`                  | `string`     | The e-mail address of the payer. Will be used to prefill the Checkin as well as on the payer's profile, if not already set. Increases the chance for [frictionless 3-D Secure 2 flow](/{{ documentation_section }}/other-features#3-d-secure-2).                                                                                                                                         |

Reviewer: The `paymentorder` and `payer` fields need to be added in the top of the table for the object hierarchy to make sense.

{% include settlement-reconciliation.md api_resource="paymentorders" %}

## 3-D Secure 2.0

When dealing with card payments, 3-D Secure authentication of the
cardholder is an essential topic. 3-D Secure 2.0 is an improved version of the
old protocol, now allowing frictionless payments where transactions can be
completed without input from the cardholder. Therefore, there are certain fields
that should be included when implementing 3-D Secure 2.0. These are listed below
in the abbreviated `payer` resource that contains payer information related to
the payment order.

*   `paymentorder.payer.email`
*   `paymentorder.payer.msisdn`
*   `paymentorder.payer.workPhoneNumber`

> *   `paymentorder.payer.homePhoneNumber`

Reviewer: This list can be removed when we have a full example, but let's add a field table below the example that highlights the 3-D Secure 2 related fields.

{:.code-header}
**Request**

```http
GET /psp/paymentorders/{{ page.payment_order_id }}/payers/ HTTP/1.1
Host: {{ page.api_host }}
Authorization: Bearer <AccessToken>
Content-Type: application/json
```

{:.code-header}
**Response**

```http
HTTP/1.1 200 OK
Content-Type: application/json

{
    "paymentorder": "/psp/paymentorders/{{ page.payment_order_id }}",
    "payer" : {
        "id": "/psp/paymentorders/{{ page.payment_order_id }}/payer",
        "reference": "reference to payer",
        "email": "email",
        "msisdn": "msisdn",
        "shippingAddress": {
            "addressee": "firstName + lastName",
            "coAddress": "coAddress",
            "streetAddress": "streetAddress",
            "zipCode": "zipCode",
            "city": "city",
            "countryCode": "countryCode"
        }
    }
}

> ```

Reviewer: Can we just show a `POST` request containing the `payer` fields, like we do for the Card example?

{% include settlement-reconciliation.md api_resource="paymentorders" %}

## 3-D Secure 2

When dealing with card payments, 3-D Secure authentication of the
cardholder is an essential topic. 3-D Secure 2 is an improved version of the
old protocol, now allowing frictionless payments where transactions can be
completed without input from the cardholder. Therefore, there are certain fields
that should be included when implementing 3-D Secure 2. These are listed below
in the abbreviated `Purchase` request. 

{:.code-header}
**Request**

```http
POST /psp/paymentorders HTTP/1.1
Host: {{ page.api_host }}
Authorization: Bearer <AccessToken>
Content-Type: application/json

{
    "paymentorder": {
        "operation": "Purchase",
        "currency": "SEK",
        "amount": 1500,
        "vatAmount": 375,
        "description": "Test Purchase",
        "generatePaymentToken": true,
        "userAgent": "Mozilla/5.0...",
        "urls": {
            "hostUrls": [ "https://example.com", "https://example.net" ]
        },

>         "payer": {

Reviewer: Thinking long about this, I don't think it's the `payer` object we should have here at all. In [the documentation](https://developer.swedbankpay.com/checkout/other-features#creating-a-payment-order) you can see that we have `riskIndicator` for `paymentorder` just like `payment`.

We can therefore just create an include from the information you now have in `payments/card/other-features.md`, which takes `api_resource` as an argument to change `paymentorder` with `payment` in the code example depending on where it's included, so we don't need to duplicate the example, field tables, etc.

            "email": "olivia.nyhuus@payex.com",
            "msisdn": "+4798765432",
            "workPhoneNumber" : "+4787654321",
            "homePhoneNumber" : "+4776543210",
            "shippingAddress": {
                "addressee": "Olivia Nyhuus",
               "streetAddress": "Saltnestoppen 43",
                "coAddress": "",
                "city": "Saltnes",
                "zipCode": "1642",
                "countryCode": "NO"
            }
        }
    }
}
```

{:.table .table-striped}
| Field                     | Type     | Description                                                                                                                                        |
| :------------------------ | :------- | :------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `paymentorder`            | `string` | {% include field-description-id.md resource="paymentorder" sub_resource="payer" %}                                                                 |

Reviewer: ```suggestion
| `paymentorder`            | `object` | {% include field-description-id.md resource="paymentorder" sub_resource="payer" %}                                                                 |
```

| `payer`                   | `object` | The payer object.                                                                                                                                  |
| └➔&nbsp;`id`              | `string` | {% include field-description-id.md resource="payer" %}                                                                                             |
| └➔&nbsp;`email`           | `string` | Payer's registered email address. The field is related to [3-D Secure 2](/{{ local_documentation_section }}/other-features#3-d-secure-2).       |
| └➔&nbsp;`msisdn`          | `string` | Payer's registered mobile phone number. The field is related to [3-D Secure 2](/{{ local_documentation_section }}/other-features#3-d-secure-2). |
| └➔&nbsp;`shippingAddress` | `object` | The shipping address object related to the `payer`. The field is related to [3-D Secure 2](/{{ local_documentation_section }}/other-features#3-d-secure-2).                                                                                                |

> | └─➔&nbsp;`addresse`       | `object` | The shipping address object related to the `payer`.                                                                                                |

Reviewer: ```suggestion
| └─➔&nbsp;`addresse`       | `string` | The name of the addressee – the receiver of the shipped goods.                                                                                                |
```
